File: src/graphs/cpp/graph/test/test_graph.cpp
#include "graph.h"
#include "gtest/gtest.h"

#include <algorithm>

template <typename T>
void assertVectorsEqual(std::vector<T> x, std::vector<T> y) {
  ASSERT_EQ(x.size(), y.size()) << "Vectors x and y are of unequal length";

  sort(x.begin(), x.end());
  sort(y.begin(), y.end());

  for (int i = 0; i < x.size(); ++i) {
    EXPECT_EQ(x[i], y[i]) << "Vectors x and y differ at index " << i;
  }
}

TEST(Vertex, EqualInt) {
  auto vertexA = Vertex<int>(1);
  auto vertexB = Vertex<int>(1);

  EXPECT_EQ(vertexA, vertexB);
}

TEST(Vertex, EqualChar) {
  auto vertexA = Vertex<char>('A');
  auto vertexB = Vertex<char>('A');

  EXPECT_EQ(vertexA, vertexB);
}

TEST(Vertex, NotEqualInt) {
  auto vertexA = Vertex<int>(1);
  auto vertexB = Vertex<int>(2);

  EXPECT_NE(vertexA, vertexB);
}

TEST(Vertex, NotEqualChar) {
  auto vertexA = Vertex<char>('A');
  auto vertexB = Vertex<char>('B');

  EXPECT_NE(vertexA, vertexB);
}

TEST(Vertex, Sort) {
  std::vector<Vertex<int>> array = {Vertex<int>(5), Vertex<int>(4),
                                    Vertex<int>(1), Vertex<int>(3),
                                    Vertex<int>(2)};
  std::vector<Vertex<int>> expectedResult = {Vertex<int>(1), Vertex<int>(2),
                                             Vertex<int>(3), Vertex<int>(4),
                                             Vertex<int>(5)};
  std::sort(array.begin(), array.end());

  assertVectorsEqual(array, expectedResult);
}

TEST(Edge, EqualInt) {
  auto edgeA = Edge<int>(Vertex<int>(1), Vertex<int>(2), 10);
  auto edgeB = Edge<int>(Vertex<int>(1), Vertex<int>(2), 10);

  EXPECT_EQ(edgeA, edgeB);
}

TEST(Edge, EqualChar) {
  auto edgeA = Edge<char>(Vertex<char>('A'), Vertex<char>('B'), 5);
  auto edgeB = Edge<char>(Vertex<char>('A'), Vertex<char>('B'), 5);

  EXPECT_EQ(edgeA, edgeB);
}

TEST(Edge, NotEqualInt) {
  auto edgeA = Edge<int>(Vertex<int>(1), Vertex<int>(2), 10);
  auto edgeB = Edge<int>(Vertex<int>(1), Vertex<int>(2), 3);

  EXPECT_NE(edgeA, edgeB);

  edgeA = Edge<int>(Vertex<int>(3), Vertex<int>(2), 10);
  edgeB = Edge<int>(Vertex<int>(1), Vertex<int>(2), 10);

  EXPECT_NE(edgeA, edgeB);
}

TEST(Edge, NotEqualChar) {
  auto edgeA = Edge<char>(Vertex<char>('A'), Vertex<char>('B'), 5);
  auto edgeB = Edge<char>(Vertex<char>('A'), Vertex<char>('C'), 5);

  EXPECT_NE(edgeA, edgeB);
}

TEST(Edge, Sort) {
  std::vector<Edge<int>> array = {Edge<int>(Vertex<int>(1), Vertex<int>(2), 10),
                                  Edge<int>(Vertex<int>(3), Vertex<int>(7), 2),
                                  Edge<int>(Vertex<int>(5), Vertex<int>(4), 7),
                                  Edge<int>(Vertex<int>(6), Vertex<int>(2), 5),
                                  Edge<int>(Vertex<int>(9), Vertex<int>(3), 8)};

  std::vector<Edge<int>> expectedResult = {
      Edge<int>(Vertex<int>(3), Vertex<int>(7), 2),
      Edge<int>(Vertex<int>(6), Vertex<int>(2), 5),
      Edge<int>(Vertex<int>(5), Vertex<int>(4), 7),
      Edge<int>(Vertex<int>(9), Vertex<int>(3), 8),
      Edge<int>(Vertex<int>(1), Vertex<int>(2), 10)};
  std::sort(array.begin(), array.end());

  assertVectorsEqual(array, expectedResult);
}

TEST(Graph, Empty) {
  Graph<int> graph;
  ASSERT_TRUE(graph.empty());
}

TEST(Graph, AddingVertices) {
  Graph<int> graph;
  graph.addVertex(Vertex<int>(1));
  graph.addVertex(Vertex<int>(2));
  graph.addVertex(Vertex<int>(3));

  ASSERT_TRUE(graph.contains(1));
  ASSERT_TRUE(graph.contains(2));
  ASSERT_TRUE(graph.contains(3));
  EXPECT_EQ(graph.size(), 3);

  assertVectorsEqual(
      graph.vertices(),
      std::vector<Vertex<int>>{Vertex<int>(1), Vertex<int>(2), Vertex<int>(3)});
}

TEST(Graph, AddingEdges) {
  Graph<int> graph;
  graph.addEdge(Vertex<int>(1), Vertex<int>(5), 10);
  graph.addEdge(Vertex<int>(1), Vertex<int>(2), 10);
  graph.addEdge(Vertex<int>(1), Vertex<int>(3), 10);
  graph.addEdge(Vertex<int>(2), Vertex<int>(2), 1);
  graph.addEdge(Vertex<int>(3), Vertex<int>(3), 2);
  graph.addEdge(Vertex<int>(4), Vertex<int>(-5), 3);
  EXPECT_EQ(graph.size(), 6);

  ASSERT_TRUE(graph.contains(1));
  ASSERT_TRUE(graph.contains(2));
  ASSERT_TRUE(graph.contains(3));
  ASSERT_TRUE(graph.contains(4));

  assertVectorsEqual(
      graph.edgesFromVertex(Vertex<int>(1)),
      std::vector<Edge<int>>{Edge<int>(Vertex<int>(1), Vertex<int>(5), 10),
                             Edge<int>(Vertex<int>(1), Vertex<int>(2), 10),
                             Edge<int>(Vertex<int>(1), Vertex<int>(3), 10)});

  assertVectorsEqual(
      graph.edgesFromVertex(Vertex<int>(2)),
      std::vector<Edge<int>>{Edge<int>(Vertex<int>(2), Vertex<int>(2), 1)});

  assertVectorsEqual(
      graph.edgesFromVertex(Vertex<int>(3)),
      std::vector<Edge<int>>{Edge<int>(Vertex<int>(3), Vertex<int>(3), 2)});

  assertVectorsEqual(
      graph.edgesFromVertex(Vertex<int>(4)),
      std::vector<Edge<int>>{Edge<int>(Vertex<int>(4), Vertex<int>(-5), 3)});
}

TEST(Graph, EdgesFromVertex) {
  Graph<int> graph;

  auto vertex = Vertex<int>(1);

  auto edgeA = Edge<int>(vertex, Vertex<int>(9), 8);
  auto edgeB = Edge<int>(vertex, Vertex<int>(4), 5);
  auto edgeC = Edge<int>(vertex, Vertex<int>(3), -5);

  graph.addEdge(edgeA);
  graph.addEdge(edgeB);
  graph.addEdge(edgeC);

  auto expectedResult = std::vector<Edge<int>>{edgeA, edgeB, edgeC};

  assertVectorsEqual(graph.edgesFromVertex(vertex), expectedResult);
}

TEST(Graph, Contains) {
  Graph<char> graph;

  auto vertex = Vertex<char>('A');
  ASSERT_FALSE(graph.contains(vertex));

  graph.addVertex(vertex);

  ASSERT_TRUE(graph.contains(vertex));
}

TEST(Graph, size) {
  Graph<int> graph;
  graph.addEdge(Vertex<int>(1), Vertex<int>(9), 10);
  graph.addEdge(Vertex<int>(5), Vertex<int>(4), 10);
  graph.addEdge(Vertex<int>(3), Vertex<int>(1), 10);
  EXPECT_EQ(graph.size(), 5);
}

File: src/graphs/python/graph/tests/test_graph.py
import unittest

import os
import sys

file_dir = os.path.dirname(os.path.dirname(__file__))
sys.path.append(file_dir + "/src")

from graph import Vertex, Edge, Graph


class TestVertex(unittest.TestCase):
    def test_equal(self):
        vertex_a = Vertex(1)
        vertex_b = Vertex(1)

        self.assertEqual(vertex_a, vertex_b)

        vertex_a = Vertex("A")
        vertex_b = Vertex("A")

        self.assertEqual(vertex_a, vertex_b)

    def test_not_equal(self):
        vertex_a = Vertex(1)
        vertex_b = Vertex(2)

        self.assertNotEqual(vertex_a, vertex_b)

        vertex_a = Vertex("A")
        vertex_b = Vertex("B")

        self.assertNotEqual(vertex_a, vertex_b)

    def test_sorting(self):
        array = [Vertex(5), Vertex(4), Vertex(1), Vertex(3), Vertex(2)]
        expected_result = [Vertex(1), Vertex(2), Vertex(3), Vertex(4), Vertex(5)]

        self.assertListEqual(sorted(array), expected_result)


class TestEdge(unittest.TestCase):
    def test_equal(self):
        edge_a = Edge(Vertex(1), Vertex(2), 10)
        edge_b = Edge(Vertex(1), Vertex(2), 10)

        self.assertEqual(edge_a, edge_b)

        edge_a = Edge(Vertex("A"), Vertex("B"), 5)
        edge_b = Edge(Vertex("A"), Vertex("B"), 5)

        self.assertEqual(edge_a, edge_b)

    def test_not_equal(self):
        edge_a = Edge(Vertex(1), Vertex(2), 10)
        edge_b = Edge(Vertex(2), Vertex(1), 10)

        self.assertNotEqual(edge_a, edge_b)

        edge_a = Edge(Vertex(1), Vertex(2), 3)
        edge_b = Edge(Vertex(2), Vertex(2), 5)

        self.assertNotEqual(edge_a, edge_b)

        edge_a = Edge(Vertex(1), Vertex(2), 10)
        edge_b = Edge(Vertex(3), Vertex(2), 10)

        self.assertNotEqual(edge_a, edge_b)

        edge_a = Edge(Vertex("A"), Vertex("B"), 5)
        edge_b = Edge(Vertex("C"), Vertex("A"), 5)

        self.assertNotEqual(edge_a, edge_b)

    def test_sorting(self):
        array = [
            Edge(Vertex(1), Vertex(9), 8),
            Edge(Vertex(7), Vertex(2), 4),
            Edge(Vertex(3), Vertex(1), -5),
            Edge(Vertex(5), Vertex(4), 5),
        ]
        expected_result = [
            Edge(Vertex(3), Vertex(1), -5),
            Edge(Vertex(7), Vertex(2), 4),
            Edge(Vertex(5), Vertex(4), 5),
            Edge(Vertex(1), Vertex(9), 8),
        ]

        self.assertListEqual(sorted(array), expected_result)


class TestGraph(unittest.TestCase):
    def test_add_edge(self):
        graph = Graph()

        self.assertDictEqual(graph._adj_dict, dict())

        edge_a = Edge(Vertex(1), Vertex(9), 8)
        edge_b = Edge(Vertex(5), Vertex(4), 5)
        edge_c = Edge(Vertex(3), Vertex(1), -5)

        graph.add_edge(edge_a)
        graph.add_edge(edge_b)
        graph.add_edge(edge_c)

        expected_result = {
            edge_a.source: [edge_a],
            edge_b.source: [edge_b],
            edge_c.source: [edge_c],
            edge_a.destination: list(),
            edge_b.destination: list(),
        }

        self.assertDictEqual(graph._adj_dict, expected_result)

    def test_add_vertex(self):
        graph = Graph()

        self.assertDictEqual(graph._adj_dict, dict())

        vertex_a = Vertex("A")
        vertex_b = Vertex("B")
        vertex_c = Vertex("C")

        graph.add_vertex(vertex_a)
        graph.add_vertex(vertex_b)
        graph.add_vertex(vertex_c)

        expected_result = {vertex_a: list(), vertex_b: list(), vertex_c: list()}

        self.assertDictEqual(graph._adj_dict, expected_result)

    def test_vertices(self):
        graph = Graph()

        vertex_a = Vertex("A")
        vertex_b = Vertex("B")
        vertex_c = Vertex("C")

        graph.add_vertex(vertex_a)
        graph.add_vertex(vertex_b)
        graph.add_vertex(vertex_c)

        expected_result = [vertex_a, vertex_b, vertex_c]

        self.assertSetEqual(set(graph.vertices()), set(expected_result))

    def test_edges(self):
        graph = Graph()

        self.assertDictEqual(graph._adj_dict, dict())

        edge_a = Edge(Vertex(1), Vertex(9), 8)
        edge_b = Edge(Vertex(5), Vertex(4), 5)
        edge_c = Edge(Vertex(3), Vertex(1), -5)

        graph.add_edge(edge_a)
        graph.add_edge(edge_b)
        graph.add_edge(edge_c)

        expected_result = [edge_a, edge_b, edge_c]

        self.assertListEqual(graph.edges(), expected_result)

    def test_edges_from_vertex(self):
        graph = Graph()

        vertex = Vertex(1)
        edge_a = Edge(vertex, Vertex(9), 8)
        edge_b = Edge(vertex, Vertex(4), 5)
        edge_c = Edge(vertex, Vertex(3), -5)

        graph.add_edge(edge_a)
        graph.add_edge(edge_b)
        graph.add_edge(edge_c)

        expected_result = [edge_a, edge_b, edge_c]

        self.assertListEqual(graph.edges_from_vertex(vertex), expected_result)

    def test_contains(self):
        graph = Graph()

        vertex = Vertex("A")

        self.assertFalse(graph.contains(vertex))
        graph.add_vertex(vertex)

        self.assertTrue(graph.contains(vertex))

    def test_connected(self):
        graph = Graph()

        edge_a = Edge(Vertex(1), Vertex(9), 8)
        edge_b = Edge(Vertex(5), Vertex(4), 5)
        edge_c = Edge(Vertex(3), Vertex(1), -5)

        graph.add_edge(edge_a)
        graph.add_edge(edge_b)
        graph.add_edge(edge_c)

        self.assertTrue(graph.connected(Vertex(1), Vertex(9)))
        self.assertTrue(graph.connected(Vertex(9), Vertex(1)))
        self.assertTrue(graph.connected(Vertex(4), Vertex(5)))
        self.assertTrue(graph.connected(Vertex(5), Vertex(4)))
        self.assertTrue(graph.connected(Vertex(1), Vertex(3)))
        self.assertTrue(graph.connected(Vertex(3), Vertex(1)))

        self.assertFalse(graph.connected(Vertex(1), Vertex(5)))
        self.assertFalse(graph.connected(Vertex(5), Vertex(1)))
        self.assertFalse(graph.connected(Vertex(1), Vertex(4)))
        self.assertFalse(graph.connected(Vertex(4), Vertex(1)))
        self.assertFalse(graph.connected(Vertex(9), Vertex(5)))
        self.assertFalse(graph.connected(Vertex(5), Vertex(9)))
        self.assertFalse(graph.connected(Vertex(9), Vertex(4)))
        self.assertFalse(graph.connected(Vertex(4), Vertex(9)))

    def test_empty(self):
        graph = Graph()
        self.assertTrue(graph.empty())

    def test_size(self):
        graph = Graph()

        self.assertTrue(graph.size() == 0)

        edge_a = Edge(Vertex(1), Vertex(9), 8)
        edge_b = Edge(Vertex(5), Vertex(4), 5)
        edge_c = Edge(Vertex(3), Vertex(1), -5)

        graph.add_edge(edge_a)
        graph.add_edge(edge_b)
        graph.add_edge(edge_c)

        self.assertTrue(graph.size() == 5)


if __name__ == "__main__":
    unittest.main()
